File: api/speech/audio/wav.ts
import { AudioFormat, AudioBuffer, SpeechError } from '../types.js';

/** WAV file utilities */
export class WavUtil {
  /** Parses WAV header and returns audio format */
  static parseHeader(buffer: Buffer): AudioFormat {
    if (buffer.toString('ascii', 0, 4) !== 'RIFF') {
      throw new SpeechError('Invalid WAV file: missing RIFF header', 'AUDIO_PARSE_ERROR');
    }
    if (buffer.toString('ascii', 8, 12) !== 'WAVE') {
      throw new SpeechError('Invalid WAV file: missing WAVE header', 'AUDIO_PARSE_ERROR');
    }

    const channels = buffer.readUInt16LE(22);
    const sampleRate = buffer.readUInt32LE(24);
    const bitDepth = buffer.readUInt16LE(34);

    return {
      sampleRate,
      channels,
      bitDepth,
      format: 'wav',
    };
  }

  /** Builds a WAV file from PCM data */
  static buildWav(pcmData: Buffer, format: AudioFormat): Buffer {
    const byteRate = format.sampleRate * format.channels * (format.bitDepth / 8);
    const blockAlign = format.channels * (format.bitDepth / 8);
    const dataSize = pcmData.length;

    const header = Buffer.alloc(44);
    header.write('RIFF', 0);
    header.writeUInt32LE(36 + dataSize, 4);
    header.write('WAVE', 8);
    header.write('fmt ', 12);
    header.writeUInt32LE(16, 16);
    header.writeUInt16LE(1, 20);
    header.writeUInt16LE(format.channels, 22);
    header.writeUInt32LE(format.sampleRate, 24);
    header.writeUInt32LE(byteRate, 28);
    header.writeUInt16LE(blockAlign, 32);
    header.writeUInt16LE(format.bitDepth, 34);
    header.write('data', 36);
    header.writeUInt32LE(dataSize, 40);

    return Buffer.concat([header, pcmData]);
  }

  /** Extracts PCM data from WAV buffer */
  static extractPcm(wavBuffer: Buffer): Buffer {
    const dataOffset = this.findDataOffset(wavBuffer);
    return wavBuffer.subarray(dataOffset);
  }

  private static findDataOffset(buffer: Buffer): number {
    for (let i = 12; i < buffer.length - 8; i++) {
      if (buffer.toString('ascii', i, i + 4) === 'data') {
        return i + 8;
      }
    }
    return 44;
  }

  /** Concatenates multiple WAV buffers (e.g. one per streamed sentence,
   * in sequence order) into a single continuous WAV file with one header.
   *
   * Naively concatenating raw WAV bytes back-to-back produces a corrupt
   * file — every chunk after the first still has its own RIFF/fmt/data
   * header embedded mid-stream, which most players either choke on or
   * read as garbage/silence. This instead parses the format from the
   * first chunk, strips every chunk down to raw PCM, concatenates the
   * PCM only, and rebuilds one correct header sized for the full thing.
   *
   * All chunks are assumed to share the same format (sample rate/channels/
   * bit depth) — true here since every chunk in a single response comes
   * from the same provider+voice call. Throws if given an empty array. */
  static concat(chunks: Buffer[]): Buffer {
    if (chunks.length === 0) {
      throw new SpeechError('Cannot concatenate zero WAV chunks', 'AUDIO_PARSE_ERROR');
    }
    if (chunks.length === 1) {
      return chunks[0];
    }

    const format = this.parseHeader(chunks[0]);
    const pcmParts = chunks.map((chunk) => this.extractPcm(chunk));
    const joinedPcm = Buffer.concat(pcmParts);

    return this.buildWav(joinedPcm, format);
  }
}

File: speech-runtime/audio/wav.py
"""WAV file utilities."""
import struct
from io import BytesIO
from typing import Tuple
import numpy as np

from runtime.exceptions import AudioError


class WavUtil:
    """WAV format utilities."""

    @staticmethod
    def parse_header(data: bytes) -> Tuple[int, int, int]:
        """Parse WAV header and return (sample_rate, channels, bit_depth)."""
        if len(data) < 44:
            raise AudioError("WAV data too short")
        if data[:4] != b"RIFF" or data[8:12] != b"WAVE":
            raise AudioError("Invalid WAV header")
        channels = struct.unpack("<H", data[22:24])[0]
        sample_rate = struct.unpack("<I", data[24:28])[0]
        bit_depth = struct.unpack("<H", data[34:36])[0]
        return sample_rate, channels, bit_depth

    @staticmethod
    def extract_pcm(data: bytes) -> bytes:
        """Extract raw PCM data from WAV."""
        # Find data chunk
        idx = 12
        while idx < len(data) - 8:
            chunk_id = data[idx:idx+4]
            chunk_size = struct.unpack("<I", data[idx+4:idx+8])[0]
            if chunk_id == b"data":
                return data[idx+8:idx+8+chunk_size]
            idx += 8 + chunk_size
        raise AudioError("WAV data chunk not found")

    @staticmethod
    def build_wav(pcm_data: bytes, sample_rate: int, channels: int = 1, bit_depth: int = 16) -> bytes:
        """Build a WAV file from PCM data."""
        byte_rate = sample_rate * channels * (bit_depth // 8)
        block_align = channels * (bit_depth // 8)
        data_size = len(pcm_data)

        header = struct.pack(
            "<4sI4s4sIHHIIHH4sI",
            b"RIFF",
            36 + data_size,
            b"WAVE",
            b"fmt ",
            16,
            1,
            channels,
            sample_rate,
            byte_rate,
            block_align,
            bit_depth,
            b"data",
            data_size,
        )
        return header + pcm_data

    @staticmethod
    def pcm_to_numpy(pcm_data: bytes, bit_depth: int = 16) -> np.ndarray:
        """Convert PCM bytes to numpy array."""
        if bit_depth == 16:
            return np.frombuffer(pcm_data, dtype=np.int16).astype(np.float32) / 32768.0
        elif bit_depth == 32:
            return np.frombuffer(pcm_data, dtype=np.int32).astype(np.float32) / 2147483648.0
        elif bit_depth == 8:
            return np.frombuffer(pcm_data, dtype=np.uint8).astype(np.float32) / 128.0 - 1.0
        else:
            raise AudioError(f"Unsupported bit depth: {bit_depth}")

    @staticmethod
    def numpy_to_pcm(audio: np.ndarray, bit_depth: int = 16) -> bytes:
        """Convert numpy array to PCM bytes."""
        if bit_depth == 16:
            audio = np.clip(audio, -1.0, 1.0)
            return (audio * 32767).astype(np.int16).tobytes()
        elif bit_depth == 32:
            audio = np.clip(audio, -1.0, 1.0)
            return (audio * 2147483647).astype(np.int32).tobytes()
        else:
            raise AudioError(f"Unsupported bit depth: {bit_depth}")
